### ISSUE TITLE
Linux editor build compiles but does not run

### ISSUE THREAD

**Issue by subject546** [AtomicGameEngine/AtomicGameEngine#282]
hi! i'm new to AtomicGameEngine but i can't get it to work propperly on my debian jessie system...

i followed the guide on from JoshEngebretson (3rd post) in this post https://github.com/AtomicGameEngine/AtomicGameEngine/issues/56

i can compile fine but when i start the atomic editor i get this error:
[Mon Aug 31 17:12:21 2015] INFO: Opened log file /home/subject546/.local/share/AtomicEditor/Logs/AtomicEditor.log
[Mon Aug 31 17:12:21 2015] INFO: Created 7 worker threads
[Mon Aug 31 17:12:21 2015] INFO: Added resource path /home/subject546/Documents/atomic game engine/AtomicGameEngine/Resources/CoreData/
[Mon Aug 31 17:12:21 2015] INFO: Added resource path /home/subject546/Documents/atomic game engine/AtomicGameEngine/Resources/EditorData/
[Mon Aug 31 17:12:21 2015] INFO: Added resource path /home/subject546/Documents/atomic game engine/AtomicGameEngine/Script/
[Mon Aug 31 17:12:21 2015] DEBUG: Skipped autoload path 'Autoload' as it does not exist, check the documentation on how to set the 'resource prefix path'
[Mon Aug 31 17:12:21 2015] INFO: Set screen mode 1024x768 windowed resizable
[Mon Aug 31 17:12:21 2015] INFO: Initialized input
[Mon Aug 31 17:12:21 2015] DEBUG: Loading resource Textures/Ramp.png
[Mon Aug 31 17:12:21 2015] DEBUG: Loading temporary resource Textures/Ramp.xml
[Mon Aug 31 17:12:21 2015] DEBUG: Loading resource Textures/Spot.png
[Mon Aug 31 17:12:21 2015] DEBUG: Loading temporary resource Textures/Spot.xml
[Mon Aug 31 17:12:21 2015] DEBUG: Loading resource Techniques/NoTexture.xml
[Mon Aug 31 17:12:21 2015] DEBUG: Loading resource RenderPaths/Forward.xml
[Mon Aug 31 17:12:21 2015] INFO: Initialized renderer
[Mon Aug 31 17:12:21 2015] INFO: Set audio mode 44100 Hz stereo interpolated
[Mon Aug 31 17:12:21 2015] INFO: Initialized engine
[Mon Aug 31 17:12:21 2015] INFO: Initialized system user interface
[Mon Aug 31 17:12:21 2015] DEBUG: Loading resource UI/DefaultStyle.xml
[Mon Aug 31 17:12:21 2015] DEBUG: Loading resource Textures/UI.png
[Mon Aug 31 17:12:21 2015] DEBUG: Loading temporary resource Textures/UI.xml
[Mon Aug 31 17:12:21 2015] DEBUG: Loading resource Fonts/Anonymous Pro.ttf
[Mon Aug 31 17:12:21 2015] DEBUG: Font face Anonymous Pro (11pt) has 624 glyphs
[Mon Aug 31 17:12:21 2015] ERROR: Could not find resource AtomicEditor/out/(Unknown Filename).js.map
(Unknown Filename) - Failed loading module: main.js - Line: -1
Error executing main.js

**Comment by JoshEngebretson** [comment]
Hello, what is happening here is that the editor scripts aren't being compiled on Linux.  

The Linux build is in our Continuous Integration system, so we are keeping it building.  However, we don't currently have the resources to ensure that it runs.  So, we update it when possible.

We are looking for someone to help with the Linux platform.

**Comment by subject546** [comment]
ah... i see i will try again in a little while ;P
i would like to help you with the linux version but i don't know if i am the right person for the job.. i'm still kindof a noobie...
anyway if there is anything i can help with please let me know :D

**Comment by CTrauma** [comment]
If you are still looking for a person to help with the Linux build I would like to throw my hat in the ring since I have a vested interest in development on Linux. C++ developer with build and release engineering experience in Linux and Windows.. Give me a shout on this thread if there is any interest, I would be more than happy to ensure that the Linux build can go toe to toe with the other platforms :).

**Comment by JoshEngebretson** [comment]
@CTrauma That would be fantastic and we can really use the help. 

The first thing I know is happening is the Editor's TypeScript isn't being compiled on Linux.  On Windows and Mac, there is a packaged node in the repo.  On Linux, since it has such great package management, I would say that this isn't necessary and we can use the system node (as a dependency).  

In Build/TypeScript there is the tsc.js compiler.  This is invoked here, to compile the editor scripts: 
https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Source/AtomicEditor/CMakeLists.txt#L60

That would be a start :)

**Comment by CTrauma** [comment]
Happy to do my part.. I'll start looking at it..

**Comment by CTrauma** [comment]
My CMake-fu is embarrassingly substandard, but I've got a fix in place for the problem and can now build and launch the editor on Ubuntu. How would you like me to provide the patch?

**Comment by JoshEngebretson** [comment]
Hey, that is awesome! :+1:   

Can you fork the repo and make a PR?  I can look over the CMake... no one really likes CMake scripting I don't think, but it generates good build output :)

**Comment by CTrauma** [comment]
Will do, I'll generate a PR.. yea.. My sentiments exactly on CMake, necessary evil, but we always end up in a fight.. I've moved most of my own stuff to SCons, but there is still plenty of 3rd party libs with cmake to deal with..

**Comment by honigbeutler123** [comment]
Is this still ongoing ? i would be happy to do something ... also i have another starting bug than the one explained in the first post.


```

[Fri Mar 11 11:41:24 2016] INFO: Opened log file /home/honigbeutler/.local/share/AtomicEditor/Logs/AtomicEditor.log
[Fri Mar 11 11:41:24 2016] INFO: Created 3 worker threads
[Fri Mar 11 11:41:24 2016] INFO: Added resource path /projects/AtomicFork/Artifacts/AtomicEditor/Resources/CoreData/
[Fri Mar 11 11:41:24 2016] INFO: Added resource path /projects/AtomicFork/Artifacts/AtomicEditor/Resources/EditorData/
[Fri Mar 11 11:41:24 2016] DEBUG: Skipped autoload path 'Autoload' as it does not exist, check the documentation on how to set the 'resource prefix path'
[Fri Mar 11 11:41:24 2016] INFO: Set screen mode 3000x1012 windowed resizable
[Fri Mar 11 11:41:24 2016] INFO: Initialized input
[Fri Mar 11 11:41:24 2016] DEBUG: Loading resource Textures/Ramp.png
[Fri Mar 11 11:41:24 2016] DEBUG: Loading temporary resource Textures/Ramp.xml
[Fri Mar 11 11:41:24 2016] DEBUG: Loading resource Textures/Spot.png
[Fri Mar 11 11:41:24 2016] DEBUG: Loading temporary resource Textures/Spot.xml
[Fri Mar 11 11:41:24 2016] DEBUG: Loading resource Techniques/NoTexture.xml
[Fri Mar 11 11:41:24 2016] DEBUG: Loading resource RenderPaths/Forward.xml
[Fri Mar 11 11:41:24 2016] INFO: Initialized renderer
[Fri Mar 11 11:41:24 2016] INFO: Set audio mode 44100 Hz stereo interpolated
[Fri Mar 11 11:41:24 2016] INFO: Initialized engine
[0311/114124:FATAL:content_main_runner.cc(711)] Check failed: base::i18n::InitializeICU(). 
#0 0x7f66a519364e <unknown>
#1 0x7f66a51a89ab <unknown>
#2 0x7f66a7d3a25a <unknown>
#3 0x7f66a50ac76f <unknown>
#4 0x7f66a50ac59f <unknown>
#5 0x7f66a505bdc8 cef_initialize
#6 0x0000016ca322 CefInitialize()
#7 0x0000016a4ee9 Atomic::WebBrowserHost::WebBrowserHost()
#8 0x0000008a7b39 AtomicEditor::AEEditorCommon::Start()
#9 0x00000089c6b6 AtomicEditor::AEEditorApp::Start()
#10 0x000000c37315 Atomic::Application::Run()
#11 0x0000008297c7 main
#12 0x7f66a17b9710 __libc_start_main
#13 0x000000861fc9 _start

```

**Comment by neikeq** [comment]
@honigbeutler123 There is [an open issue](https://bitbucket.org/chromiumembedded/cef/issues/1198) about that in the cef repo.
Copying _icudtl.dat_, _natives_blob.bin_ and _snapshot_blob.bin_ to _Artifacts/AtomicEditor_ solves the problem here.

**Comment by bdero** [comment]
I've been trying to debug a runtime issue for the past couple of days which is not the same, but definitely holds true to the title of this issue:

```
[Wed May  4 23:06:04 2016] INFO: Initialized renderer
[Wed May  4 23:06:04 2016] INFO: Set audio mode 44100 Hz stereo interpolated
[Wed May  4 23:06:04 2016] INFO: Initialized engine
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_S_construct null not valid
zsh: abort (core dumped)  ./Artifacts/AtomicEditor/AtomicEditor
```

It seems I'm hitting up against a condition which is causing some string to not be constructed properly (null).

Here's a gdb session with the [backtrace (with debug symbols)](https://gist.github.com/bdero/8f27d3f5bfa5d6a0b1923c1b68042147#file-atomic-gdb-L48-L77) after the crash: https://gist.github.com/bdero/8f27d3f5bfa5d6a0b1923c1b68042147

The deepest line in the trace with a debug symbol worth looking at is the call to [cef_initialize()](https://github.com/AtomicGameEngine/CEF3Binaries/blob/a74fdafc2f42f4349d777e1a1fb20f63e848da23/Linux/libcef_dll/wrapper/libcef_dll_wrapper.cc#L166-L170)

Unfortunately, it looks like the issue is happening deep within CEF, and the debug builds available on [cefbuilds.com](https://cefbuilds.com) revealed symbols to a bunch of files that were not in the CEF source tree (https://bitbucket.org/chromiumembedded/cef/src).

I've tried a variety of different CEF builds from https://cefbuilds.com. I also tried my hand at building CEF, but after [bending it to shape my environment](https://github.com/bdero/depot_tools/commit/685577439cbf9cb8c660e3da39bdcbb64c197c95) among other things, I decided to steer clear once I reached a point where the build tool started hanging with few leads.
I also tried compiling Atomic using both Clang and GCC.

## TL;DR

Is anyone able to get working fresh Linux builds? If so, can you let me know what commit hash you're building against for both the AtomicGameEngine repo and the Submodules/CEF/ repo?

I would also really appreciate if anyone with a working linux build could paste their output of `cat /proc/version` and `lsb_release -a`, which may provide some clues and will hopefully lead to a specific known working distro to work off of.

My output is as follows:

```bash
$ cat /proc/version 
Linux version 4.5.1-1-ARCH (builduser@tobias) (gcc version 5.3.0 (GCC) ) #1 SMP PREEMPT Thu Apr 14 19:19:32 CEST 2016
$ lsb_release -a
LSB Version:	1.4
Distributor ID:	Arch
Description:	Arch Linux
Release:	rolling
Codename:	n/a
```
I've also built a vagrant configuration for running Atomic builds against Ubuntu Xenial - which reliably reproduces the same issue (it's hopefully easy to use and should reliably produce broken Linux builds regardless of the host OS - Linux, OSX, or Windows): https://github.com/bdero/atomic-vagrant

**Comment by bdero** [comment]
I've now reproduced the problem with multiple CEF builds on Ubuntu Xenial (16.04), Ubuntu Trusty (15.04), and Arch Linux.
I've also improved the Vagrantfile so that (hopefully) anyone can reproduce the problem pretty easily.

![Screenshot](http://i.imgur.com/pEKA3LQ.png)

> VM on the left is me reproducing the problem on an Ubuntu Trusty guest; terminal emulator on the right is me reproducing the problem locally (Arch Linux).

Things to try from here:
- I haven't tried any earlier builds within the [CEF3Binaries repo](https://github.com/AtomicGameEngine/CEF3Binaries). I'll be giving the previous few hashes a shot once I have some time to spare again.
- Get a working build of CEF going in their recommended environment with debug enabled - and then be able to make sense of the debug symbols when stepping through with gdb.

**Comment by CTrauma** [comment]
I'm seeing this too as well as @dragonCASTjosh, going to try to have a deeper look at this since seems to be effecting multiple people.

**Comment by CTrauma** [comment]
I believe the error is here:

https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Source/AtomicWebView/WebBrowserHost.cpp#L131

There are only 2 args but the array is of size three with the last arg being empty which cef is trying to read.. changing the '3's to '2's should do the trick.. I'll submit a PR asap..
